### PR TITLE
Lazy loading fix

### PR DESCRIPTION
**Pull request by mdbenjam**
Connects: #3214 

We currently load VACOLS data lazily on writes as well as reads. After this change we will only load VACOLS data on reads. We should eventually turn on the commented out error that is raised when you trigger a VACOLS load after setting VACOLS values. This is to reduce the number of calls we make to VACOLS. Unfortunately, both hearings and reader set VACOLS values in the appeal, and then load others. We should consolidate these calls. There are two tickets for this: https://github.com/department-of-veterans-affairs/caseflow/issues/3213 and https://github.com/department-of-veterans-affairs/caseflow/issues/3241

**Test Plan**
- [ ] Ensure all the tests pass.

### CODE REVIEW

File: app/models/concerns/associated_vacols_model.rb [hunk 1]
```diff
@@ -13,18 +13,28 @@ module ClassMethods
...
       fields.each do |field|
         define_method field do
-          check_and_load_vacols_data!
+          check_and_load_vacols_data! unless field_set?(field)
           instance_variable_get("@#{field}".to_sym)
         end
 
         define_method "#{field}=" do |value|
-          check_and_load_vacols_data!
+          # @vacols_load_status = :disabled
```
Comment (mdbenjam) [inline app/models/concerns/associated_vacols_model.rb]: When we're ready to start throwing errors. Not sure if we actually want to do this, I worry it's going to be impossible to catch some of this in dev mode. But I guess we'll get it in UAT.

File: app/models/concerns/associated_vacols_model.rb [hunk 2]
```diff
@@ -35,6 +45,7 @@ def assign_from_vacols(values)
   end
 
   def check_and_load_vacols_data!
+    # raise LazyLoadingDisabledError if @vacols_load_status == :disabled
```
Comment (mdbenjam) [inline app/models/concerns/associated_vacols_model.rb]: Also needed when we want to turn on the errors.

File: lib/generators/appeal.rb [hunk 3]
```diff
@@ -172,9 +172,9 @@ def build(attrs = {})
       inaccessible = attrs.delete(:inaccessible)
 
       appeal = Appeal.find_or_initialize_by(vacols_id: attrs[:vacols_id])
-      appeal.attributes = attrs
```
Comment (mdbenjam) [inline lib/generators/appeal.rb]: We shouldn't be setting properties directly. We should go through the normal flow.

File: spec/models/appeal_spec.rb [hunk 4]
```diff
@@ -14,7 +14,8 @@
       video_hearing_requested: video_hearing_requested,
       appellant_first_name: "Joe",
       appellant_middle_initial: "E",
-      appellant_last_name: "Tester"
+      appellant_last_name: "Tester",
+      decision_date: nil
```
Comment (mdbenjam) [inline spec/models/appeal_spec.rb]: These generator changes for some reason required us to specifically nil out fields that need to be nil in the test if they're defined in the `vacols_record_default_attrs` in `generators/appeal.rb`

**Comment by NickHeiner** [comment]
As a reminder, since this PR interacts with VACOLS, please get two people to approve it that part of the code. Thanks!

**Comment by mdbenjam** [comment]
@NickHeiner I think that's a good idea, but I thought our team norms only asked for 2 reviewers on code that writes to VACOLS.

File: app/models/concerns/associated_vacols_model.rb [hunk 5]
```diff
@@ -13,18 +13,28 @@ module ClassMethods
...
       end
     end
   end
 
+  def field_set?(field)
+    @set_fields && @set_fields[field.to_sym]
+  end
+
+  def mark_field_is_set(field)
+    @set_fields = {} if !@set_fields
```
Comment (joofsh) [inline app/models/concerns/associated_vacols_model.rb]: the conditional check here and line 30 should be avoidable by creating a `set_fields` method that memoizes the value.
``` 
def set_fields
  @set_fields ||= {}
end
```

Ideally you could just do this in a `def initialize` but ruby/rails/concerns makes this too hard to be worth it

File: app/models/concerns/associated_vacols_model.rb [hunk 6]
```diff
@@ -12,17 +12,34 @@ module ClassMethods
...
       end
     end
+
+    def vacols_field?(field)
+      @vacols_fields && @vacols_fields[field]
+    end
+  end
+
+  def field_set?(field)
+    @set_fields && @set_fields[field]
```
Comment (joofsh) [inline app/models/concerns/associated_vacols_model.rb]: we should create methods that memoize these.
``` 
def set_fields
  @set_fields ||= {}
end
```

File: app/models/concerns/associated_vacols_model.rb [hunk 7]
```diff
@@ -35,6 +55,14 @@ def assign_from_vacols(values)
   end
 
   def check_and_load_vacols_data!
+    # When we're ready to turn the error on, replace the if below with the
+    # commented out error.
+    # raise LazyLoadingDisabledError if @vacols_load_status == :disabled
+    if @vacols_load_status == :disabled
```
Comment (joofsh) [inline app/models/concerns/associated_vacols_model.rb]: can we add a comment describing the `vacols_load_status` and possible states? It's hard to follow the logic.

File: app/models/concerns/associated_vacols_model.rb [hunk 8]
```diff
@@ -12,17 +12,37 @@ module ClassMethods
...
+    def vacols_field?(field)
+      vacols_fields[field]
+    end
+  end
+
+  def field_set?(field)
+    set_fields[field]
+  end
+
+  def mark_field_is_set(field)
```
Comment (joofsh) [inline app/models/concerns/associated_vacols_model.rb]: `mark_field_as_set`?

File: app/models/concerns/associated_vacols_model.rb [hunk 9]
```diff
@@ -12,17 +12,37 @@ module ClassMethods
     # do subsequent VACOLS DB lookups
     def vacols_attr_accessor(*fields)
       fields.each do |field|
+        @vacols_fields = {} if !@vacols_fields
```
Comment (joofsh) [inline app/models/concerns/associated_vacols_model.rb]: `@vacols_fields ||= {}` would be slightly cleaner. And I think you could move this up to before line 14 (no need to call this each loop through a list of fields)

Comment (mdbenjam) [reply app/models/concerns/associated_vacols_model.rb]: Ha I made a method `vacols_fields` to solve this problem then I forgot to call it...
